clippy_lints/src/needless_cow.rs

```diff
@@ -0,0 +1,384 @@
+use clippy_utils::diagnostics::span_lint_hir_and_then;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::{fn_has_unsatisfiable_preds, is_diag_trait_item};
+use itertools::Itertools;
+use rustc_const_eval::interpret::{AllocId, AllocRange, ConstValue, GlobalAlloc};
+use rustc_data_structures::fx::{FxIndexMap, IndexEntry};
+use rustc_hir::def_id::LocalDefId;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::{
+    self, AggregateKind, ConstantKind, Location, Operand, Place, Rvalue, SourceInfo, Statement, StatementKind,
+    TerminatorKind,
+};
+use rustc_middle::ty::TyCtxt;
+use rustc_mir_dataflow::lattice::FlatSet;
+use rustc_mir_dataflow::value_analysis::{Map, State, ValueAnalysis, ValueAnalysisWrapper, ValueOrPlace};
+use rustc_mir_dataflow::{Analysis, Results, ResultsVisitor};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, Span};
+use rustc_target::abi::{FieldIdx, Size};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for locals which are assigned a `Cow`, yet the `Borrowed` variant is only ever used for
+    /// empty string slices.
+    ///
+    /// ### Why is this bad?
+    /// An empty string does not allocate any memory. This can use `String` instead, with the
+    /// `Borrowed` constructor substituted for `String::new`.
+    ///
+    /// ### Known problems
+    /// If the `Cow` is passed to a function expecting `Cow`, or is mutably borrowed, this will
+    /// still lint.
+    ///
+    /// ### Example
+    /// ```rust
+    /// # use std::borrow::Cow;
+    ///
+    /// # let my_owned_string = String::new();
+    /// # let something = true;
+    /// # let other_thing = true;
+    ///
+    /// let _ = if something {
+    ///     Cow::Owned(my_owned_string);
+    /// } else if other_thing {
+    ///     Cow::Owned(String::new());
+    /// } else {
+    ///     Cow::Borrowed("");
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # use std::borrow::Cow;
+    ///
+    /// # let my_owned_string = String::new();
+    /// # let something = true;
+    /// # let other_thing = true;
+    ///
+    /// let _ = if something {
+    ///     my_owned_string;
+    /// } else if other_thing {
+    ///     String::new();
+    /// } else {
+    ///     String::new();
+    /// };
+    /// ```
+    #[clippy::version = "1.73.0"]
+    pub NEEDLESS_COW,
+    nursery,
+    "usage of `Cow` with only empty strings in `Borrow` variant"
+}
+impl_lint_pass!(NeedlessCow => [NEEDLESS_COW]);
+
+pub struct NeedlessCow;
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessCow {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // Probably unnecessary nowadays considering dummy MIR is created for these functions.
```

Comment:
I'd like to know whether this is the case. This is the code I was referencing: https://github.com/rust-lang/rust/blob/d7e751006cb3691d1384b74196a9cb45447acfa8/compiler/rustc_mir_transform/src/lib.rs#L381-L423